skbeam/core/correlation.py

```diff
@@ -842,3 +850,237 @@ def one_time_from_two_time(two_time_corr):
+                'regular' : divide by pixel number
+                'symavg' : use symmetric averaging or not (can be overridden by
+                    calling routine)
+
+            axes : the axes to perform the fft on can either be a 2tuple or 1
+                number.  The former computes a 2D correlation and latter 1D
+                correlation.
+                (defaults to (-2, -1) if None)
+
+        '''
+        if axes is None:
+            if len(shape) == 2:
+                axes = (-2, -1)
+            elif len(shape) == 1:
+                axes = (-1,)
+
+        if normalization is None:
+            normalization = ['symavg', 'regular']
+        elif type(normalization) is not list:
+            normalization = list([normalization])
+
+        self.axes = axes
+        self.normalization = normalization
+
+        if mask is None:
+            mask = np.ones(shape)
+
+        # the IDs for the image, called mask
+        self.mask = mask
+        # initialize all the masks for the correlation
+
+        # not really clean, but making a list of arrays holding the masks
+        # for each id. Ideally, mask is binary so this is one element
+        # to quickly index original images
+        self.pxlsts = list()
+        self.submasks = list()
+        # to quickly index the sub images
+        self.subpxlsts = list()
+        # the temporary images (double the size for the cross correlation)
+        self.tmpimgs = list()
+        self.ids = np.sort(np.unique(mask))
+        # remove the zero since we ignore, but only if it is there (sometimes
+        # may not be)
+        if self.ids[0] == 0:
+            self.ids = self.ids[1:]
+
+        self.nids = len(self.ids)
+        self.maskcorrs = list()
+        # regions where the correlations are not zero
+        self.pxlst_maskcorrs = list()
+
+        # basically saving bunch of mask related stuff like indexing etc, just
+        # to save some time when actually computing the cross correlations
+        for idno in self.ids:
+            masktmp = (mask == idno)
+            self.pxlsts.append(np.where(masktmp.ravel() == 1)[0])
+
+            # this could be replaced by skimage cropping and padding
+            submasktmp = _crop_from_mask(masktmp)
+            submask = _expand_image(submasktmp)
+
+            tmpimg = np.zeros_like(submask)
+
+            self.submasks.append(submask)
+            self.subpxlsts.append(np.where(submask.ravel() == 1)[0])
+            self.tmpimgs.append(tmpimg)
+            maskcorr = _cross_corr(submask)
+            # quick fix for finite numbers should be integer so
+            # choose some small value to threshold
+            maskcorr *= maskcorr > .5
+            self.maskcorrs.append(maskcorr)
+            self.pxlst_maskcorrs.append(maskcorr > 0)
+
+    def __call__(self, img1, img2=None, normalization=None, PF=False,
+                 return_indices=False):
+        ''' Run the analysis on an image, or two
+
```

Comment:
Also, see the [notebook](https://github.com/ordirules/notebooks/blob/master/CrossCorrelator.ipynb) referenced above, it contains sample usage cases with figures.